### PR TITLE
[new] Add FormChimp w/ git auto-update via single package.json

### LINKED ISSUE #11510: [Request] Add FormChimp
**Library name:** FormChimp
**Git repository url:** https://github.com/fabioquarantini/formchimp
**License:** MIT
**Official homepage:** http://www.fabioquarantini.com/formchimp/

=====================

For beginners who is interested in picking up this issue:

In this issue, we are going to host a new web front-end library on CDNJS, so that web developer can easily use it on our free CDN without additional download/upload process 🚀

If you'll like to work locally, I'll suggest you start pulling cdnjs repository with sparseCheckout: https://github.com/cdnjs/cdnjs/blob/master/documents/sparseCheckout.md, since there are so many files in this repo, without sparseCheckout, the whole process will be slow and will use more than 80GB disk space, using git command line can learn more skills and experience, if you don't like that, instead, you can also just do it on GitHub, it's more convenient, but it may be a little bit harder to fix problems on GitHub gui.

Here are the docs about how to add a library:
 - https://github.com/cdnjs/cdnjs/blob/master/CONTRIBUTING.md#d-adding-a-new-library-with-its-assets
 - https://github.com/cdnjs/cdnjs/blob/master/CONTRIBUTING.md#e-adding-a-new-library-by-a-single-packagejson

We'll also ask to setup auto-update config so that the added libraries can be updated to the latest version easily:
 - https://github.com/cdnjs/cdnjs/blob/master/documents/autoupdate.md

For beginners who is interested in join and help add this library, please feel free to tell me what kind of help you'll need, we'll help you send the valid pull request, thanks 😄 

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/46871519-request-add-formchimp?utm_campaign=plugin&utm_content=tracker%2F32893&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F32893&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by arghlex**
close #11510, cc @fabioquarantini

Pull request for issue: #11510

Checklist for **Pull request** or **lib adding request issue** follows the conventions.

Note that if you are using a distribution purpose repository/package, please also provide the url and other related info like popularity of the source code repo/package.

# Profile of the lib
 * Git repository (required): https://github.com/fabioquarantini/formchimp
 * Official website (optional, not the repository): http://www.fabioquarantini.com/formchimp/
 * NPM package url (optional):
 * License and its reference:
 * GitHub / Bitbucket popularity (required):
   - Count of watchers: 4
   - Count of stars: 83
   - Count of forks: 17
 * NPM download stats (optional):
   - Downloads in the last day:
   - Downloads in the last week:
   - Downloads in the last month:

# Essential checklist
 * [ ] I'm the author of this library
   * [ ] I would like to add link to the page of this library on CDNJS on website / readme
 * [x] This lib was not found on cdnjs repo
 * [x] No already exist / duplicated issue and PR
 * [ ] The lib has notable popularity
   * [ ] More than 200 [Stars / Watchers / Forks] on [GitHub / Bitbucket]
   * [ ] More than 800 downloads stats per month on npm registry
 * [ ] Project has public repository on famous online hosting platform (or been hosted on npm)

# Auto-update checklist
 * [x] Has valid tags for each versions (for git auto-update)
 * [x] Auto-update setup
 * [x] Auto-update target/source is valid.
 * [x] Auto-update filemap is correct.

# Git commit checklist
 * [x] The first line of commit message is less then 50 chars, be clean and clear, easy to understand.
 * [x] The parent of the commit(s) in the PR is not old than 3 days.
 * [x] Pull request is sending from a non-master branch with meaningful name.
 * [x] Separate unrelated changes into different commits.
 * [x] Use rebase to squash/fixup dummy/unnecessary commits into only one commit.
 * [x] Close corresponding issue in commit message
 * [x] Mention related issue(s), people in commit message, comment.

### CODE REVIEW

**Comment by arghlex** [comment]
I did this in GitHub first but had issues so downloaded the repo and ran fixFormat.js then ran npm test which fixed the formatting issues but still get one error. 

> FormChimp: filename from package.json exists and is a file (not a directory)
>./ajax/libs/FormChimp/1.2.1/jquery.formchimp.min.js does not exist but is referenced in package.json! // valid-packages-test.js:122

Not sure how to fix this, can anyone help?

File: ajax/libs/FormChimp/package.json [hunk 1]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "FormChimp",
+  "filename": "jquery.formchimp.min.js",
+  "title": "FormChimp",
+  "description": "A customizable MailChimp ajax plugin for jQuery",
+  "version": "1.2.1",
```
Comment (sufuf3) [inline ajax/libs/FormChimp/package.json]: Let's remove `"version": "1.2.1",` because we add this lib via single package.json
Thank you.

**Comment by PeterDaveHello** [comment]
@arghlex just simply remove the version info will help!

File: ajax/libs/FormChimp/package.json [hunk 2]
```diff
@@ -0,0 +1,40 @@
+{
+  "name": "FormChimp",
+  "filename": "jquery.formchimp.min.js",
+  "title": "FormChimp",
```
Comment (sufuf3) [inline ajax/libs/FormChimp/package.json]: Let's remove `"title": "FormChimp",` this field is not necessary field.
Thank you.

File: ajax/libs/FormChimp/package.json [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    "MailChimp",
+    "jQuery plugin",
+    "ajax"
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/fabioquarantini/formchimp.git"
+  },
+  "licenses":	{
+    "type": "MIT",
```
Comment (sufuf3) [inline ajax/libs/FormChimp/package.json]: Let's use the CDNJS's format.
Please modify from A to B.

-A 
```js
  "licenses":	{
    "type": "MIT",
    "url": "http://opensource.org/licenses/MIT"
  },
```

- B
```js
"license": "MIT",
```

File: ajax/libs/FormChimp/package.json [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+  },
+  "autoupdate": {
+    "source": "git",
+    "target": "https://github.com/fabioquarantini/formchimp.git",
+    "fileMap": [
+      {
+        "basePath": "",
+        "files": [
+          "jquery.formchimp.min.js",
+          "jquery.formchimp.js"
```
Comment (sufuf3) [inline ajax/libs/FormChimp/package.json]: We usually use [node-glob](https://github.com/isaacs/node-glob#glob-primer) which is the pattern mechanism to get the files.
Let's revise from A to B.

- A
```js
          "jquery.formchimp.min.js",
          "jquery.formchimp.js"
```

- B
```js
          "jquery.formchimp*.+(js|css|map)"
```

Thank you.

**Comment by arghlex** [comment]
@PeterDaveHello @sufuf3 thanks for your help, it's my first contribution to cdnjs! I will amend the file with your advice and commit soon.

**Comment by sufuf3** [comment]
@arghlex May I ask you how do you send this pull request? via GitHub website or use git command?
I think the commit should squash to one commit, if you use git command, we can help you how to do this.
But if you use GitHub website, it can't squash the commit. I can help you squash to one commit.

Thank you.

**Comment by arghlex** [comment]
@sufuf3 I used the git command in terminal on mac osx for most recent changes.

File: ajax/libs/FormChimp/package.json [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    "url": "https://github.com/fabioquarantini/formchimp.git"
+  },
+  "autoupdate": {
+    "source": "git",
+    "target": "https://github.com/fabioquarantini/formchimp.git",
+    "fileMap": [
+      {
+        "basePath": "",
+        "files": [
+          "jquery.formchimp*.+(js|css|map)"
```
Comment (sufuf3) [inline ajax/libs/FormChimp/package.json]: Sorry for carelessness, 
In this case, we need to use `"jquery.formchimp!(*-[0-9])*.+(js|css|map)"` to ignore `jquery.formchimp.min-0.1.0.js` in v1.0.0.
Thank you.

**Comment by sufuf3** [comment]
Oh, Just need to rebase to the master, then it will pass CI test.
1. `git checkout master`
2. `git pull origin master`
3. `git checkout FormChimp`
4. `git rebase master`
5. `git push your_remote_name FormChimp -f`
😃

**Comment by arghlex** [comment]
@sufuf3 Sorry, I followed your steps but the CI didn't like it again. I'm not really sure what's going wrong?

**Comment by sufuf3** [comment]
The CI said `this pull request is a little bit old`
Would you try to squash to one commit again?
And then rebase to the master again.

If CI said it's still wrong, let's see the `git log`.

Thank you.

**Comment by sufuf3** [comment]
@arghlex May I help you to fix this PR, let this branch rebase to the latest master?
I think you merge the master into FormChimp branch, not [rebase](https://git-scm.com/docs/git-rebase). 
It's not a big problem, I can help you let this PR better.
Thank you.

**Comment by arghlex** [comment]
@sufuf3 Yes please do, sorry!

**Comment by sufuf3** [comment]
@arghlex It's OK.
Here is a tutorial you could try locally: https://www.digitalocean.com/community/tutorials/how-to-create-a-pull-request-on-github
Hope this can help you. :)
Thanks for your contribution. 😃

File: ajax/libs/FormChimp/package.json [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    "name": "Fabio Quarantini",
+    "url": "http://www.fabioquarantini.com",
+    "twitter": "@febba"
+  },
+  "homepage": "http://www.fabioquarantini.com/formchimp/",
+  "keywords": [
+    "FormChimp",
+    "MailChimp",
+    "jQuery plugin",
+    "ajax"
```
Comment (extend1994) [inline ajax/libs/FormChimp/package.json]: Let's add `form` and `newsletter` like https://github.com/fabioquarantini/formchimp/blob/1.2.1/bower.json#L15 and https://github.com/fabioquarantini/formchimp/blob/1.2.1/bower.json#L17

**Comment by sufuf3** [comment]
@arghlex Would you add `form` and `newsletter` into keywords?
And about jquery.formchimp.min-0.1.0.js, I can help you manually add.

I think this PR almost finish, except for this part.
Thank you.

**Comment by mercyg** [comment]
@sufuf3 doesn't this library need to have more popularity on github and npm ?

**Comment by PeterDaveHello** [comment]
@mercyg we also promote some library without enough popularity but with good versioning, documentation, file/directory structure, so no worries 😄

**Comment by sufuf3** [comment]
@arghlex 
thanks for contribution.
Would you like to work on this and help improve this PR?
If you have any question, please let us know.

Many thanks.

**Comment by sufuf3** [comment]
@arghlex
thanks for contribution.
Would you mind working on this and helping improve this PR?
If you have any question, please let us know.

Many thanks.

**Comment by sashberd** [comment]
@sufuf3 I would like to continue this PR but I cannot do rebase to latest master. It provides me a lot of errors of pathing and merging. Could you help me to rebase to latest commit and I will continue from there? Thanks

**Comment by sufuf3** [comment]
Ummm @PeterDaveHello This issue is for the beginner. However, there is no response from the First-time contributor since 14 Aug 2017.
Should this PR be helped by @sashberd?
Or maybe we can close this PR and let the beginner try the issue?
What do you think?

Many thanks.

**Comment by PeterDaveHello** [comment]
@sufuf3 good point, maybe just close it ;)

**Comment by PeterDaveHello** [comment]
@sufuf3 feel free to close it 👍

**Comment by sufuf3** [comment]
Closed due to no response from the First-time contributor since 14 Aug 2017. 
cc @PeterDaveHello

**Pull request closed without merging**